### DISCUSSION TITLE
npm install 404 for packages in private repository RENOVATE_DETECT_GLOBAL_MANAGER_CONFIG

### DISCUSSION THREAD

**Discussion by reece-white** [renovatebot/renovate#26563] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Latest renovate/renovate docker image

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I have not seen this working

### Describe the problem

When running `renovate/renovate` latest docker image using an npm workspaces monorepo, install package fails when package.json contains private npm package.

Environment variables:
```
        RENOVATE_TMPDIR = "${env.WORKSPACE}/.tmp"
        RENOVATE_BINARY_SOURCE = "global"
        RENOVATE_DETECT_GLOBAL_MANAGER_CONFIG = true
        RENOVATE_SKIP_INSTALLS = false
        RENOVATE_CONFIG_FILE = "${env.WORKSPACE}/renovate.json"
        RENOVATE_ONBOARDING = false
        RENOVATE_NPMRC_MERGE = false
```

Unable to provide reproduction case as it requires a private npm package and token for that repository.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: File config
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "extends": [
           "config:recommended",
           "monorepo:lerna",
           "packages:eslint",
           "packages:jsTest",
           "packages:react"
         ],
         "enabledManagers": ["npm"],
         "rangeStrategy": "bump",
         "prConcurrentLimit": 5,
         "commitMessagePrefix": "[PATCH] chore",
         "constraintsFiltering": "strict",
         "constraints": {"node": "18.17.1"},
         "packageRules": [{"excludePackagePrefixes": ["@sky-uk-ott"]}]
       }
DEBUG: CLI config
       "config": {
         "repositories": ["sky-uk/core-video-sdk-js"],
         "logFile": "renovate.log",
         "logFileLevel": "debug"
       }
DEBUG: Env config
       "config": {
         "hostRules": [],
         "detectGlobalManagerConfig": true,
         "binarySource": "global",
         "onboarding": false,
         "token": "***********",
         "npmrcMerge": false,
         "skipInstalls": false
       }
DEBUG: Combined config
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "extends": [
           "config:recommended",
           "monorepo:lerna",
           "packages:eslint",
           "packages:jsTest",
           "packages:react"
         ],
         "enabledManagers": ["npm"],
         "rangeStrategy": "bump",
         "prConcurrentLimit": 5,
         "commitMessagePrefix": "[PATCH] chore",
         "constraintsFiltering": "strict",
         "constraints": {"node": "18.17.1"},
         "packageRules": [{"excludePackagePrefixes": ["@sky-uk-ott"]}],
         "hostRules": [],
         "detectGlobalManagerConfig": true,
         "binarySource": "global",
         "onboarding": false,
         "token": "***********",
         "npmrcMerge": false,
         "skipInstalls": false,
         "repositories": ["sky-uk/core-video-sdk-js"],
         "logFile": "renovate.log",
         "logFileLevel": "debug"
       }
DEBUG: Detecting global manager config
DEBUG: Detected /home/ubuntu/.npmrc and adding it to global config
DEBUG: Global manager config
       "config": {"npmrc": "***********", "npmrcMerge": true}
DEBUG: Enabling forkProcessing while in non-autodiscover mode
DEBUG: Found valid git version: 2.43.0
DEBUG: Setting global hostRules
...
DEBUG: Using personal access token for git init (repository=sky-uk/core-video-sdk-js)
DEBUG: Resetting npmrc (repository=sky-uk/core-video-sdk-js)
DEBUG: Setting npmrc (repository=sky-uk/core-video-sdk-js)
...
DEBUG: npmrc file path/.npmrc found in repository
...
DEBUG: Generating package-lock.json for path/path (repository=, branch=renovate/npm-dompurify-vulnerability)
DEBUG: Spawning npm install to create path/path/package-lock.json (repository=, branch=renovate/npm-dompurify-vulnerability)
DEBUG: Using npm constraint <9 for lockfileVersion=2 (repository=, branch=renovate/npm-dompurify-vulnerability)
DEBUG: Performing node_modules install (repository=, branch=renovate/npm-dompurify-vulnerability)
DEBUG: Performing lockfileUpdate (npm) (repository=, branch=renovate/npm-dompurify-vulnerability)
DEBUG: Executing command (repository=, branch=renovate/npm-dompurify-vulnerability)
       "command": "npm install --no-audit --ignore-scripts dompurify@2.0.17"
DEBUG: rawExec err (repository=, branch=renovate/npm-dompurify-vulnerability)
       "err": {
         "cmd": "/bin/sh -c npm install --no-audit --ignore-scripts dompurify@2.0.17",
...
npm ERR! code E404
npm ERR! 404 Not Found - GET https://**redacted**@some-repository/some-package/-/some-package-2.20.3.tgz - Not found
npm ERR! 404
npm ERR! 404  '@some-repository/some-package@https://**redacted**@some-repository/some-package/-/some-package-2.20.3.tgz' is not in this registry.
npm ERR! 404
npm ERR! 404 Note that you can also install from a\nnpm ERR! 404 tarball, folder, http url, or git url.
```
</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by reece-white** [comment]
Just to add to this, renovate seems to create an `.npmrc` file in a monorepo package based on log line `DEBUG: npmrc file path/package/.npmrc found in repository (repository=, branch=renovate/npm-dompurify-vulnerability)` but it is later ignored `WARN ignoring workspace config at path-to-monorepo-package/.npmrc`

**Comment by reece-white** [comment]
Based on https://github.com/npm/cli/issues/4605 renovate should only be creating a `.npmrc` at the root of the monorepo and not in individual packages.

**Comment by rarkins** [comment]
Why did you set npmrcMerge=false? It's overridden by the global manager config detection

**Reply by reece-white** [reply]
@rarkins Based on the log lines I can see this happens. I was trying to see if it would cause any positive changes but appears not.

**Comment by rarkins** [comment]
@reece-white why did you raise both this and #26562?

**Reply by reece-white** [reply]
@rarkins They are slightly different configurations, and both didn't work.

**Reply by rarkins** [reply]
We don't have time to troubleshoot each way you attempted. Please pick the approach you prefer, and close the other question. Alternatively, close both and open a _config question_ describing what you are trying to achieve and asking the best way to achieve it

**Reply by reece-white** [reply]
I've now closed https://github.com/renovatebot/renovate/discussions/26562, but would appreciate it if we could get RENOVATE_DETECT_GLOBAL_MANAGER_CONFIG working for npm workspaces

**Reply by rarkins** [reply]
Out of interest, why is that file-based approach your preferred instead of configuring npmrc in renovate config directly? Or are either ok?

And so you expect to have npmrc files in the repos you run renovate against?

**Reply by reece-white** [reply]
The file based approach is preferred as I don't want to commit the npm token to the repository which is where the `renovate.json` is stored. The `.npmrc` file comes from the CI environment so it de-risks the solution.

**Reply by reece-white** [reply]
It's worth noting that it isn't just the file based approach that is causing the issue in the npm workspace. Here's the renovate log snippets from setting `RENOVATE_NPM_TOKEN` as an environment variable we see:
```
npm WARN ignoring workspace config at /path/to/package/.npmrc
```

The issue seems to be that when renovate runs `npm install` it creates a new `.npmrc` file based on the config renovate is provided in the `/path/to/package/.npmrc` but `npm` ignores individual `.npmrc` in favour of one at the workspace root, which doesn't exist because renovate didn't create one.

```
DEBUG: Writing updated .npmrc file to /path/to/package/.npmrc (repository=, branch=renovate/swc-monorepo)
```

**Reply by rarkins** [reply]
That part we can consider as a bug, but we'd need a reproduction repo to be able to debug and fix it. Something like:
- monorepo with root and one package
- npmrc contents configure in config

**Reply by reece-white** [reply]
I'll try and get something together for this, but It'll probably require some work from the investigator to create a private npm package and publish it then use it inside the mock repository that I provide.

**Reply by rarkins** [reply]
I think the bug should be independent of whether the package is private or not. Renovate writes the npmrc regardless
